toolsrc/src/vcpkg/versions.cpp

```diff
@@ -0,0 +1,252 @@
+            /*vcpkg::Checks::exit_with_message(
+                VCPKG_LINE_INFO, "Unsatisfied version requirements for package: %s", package_name);*/
+        }
+    }
+
+    return ComputedVersions{computed_versions, baseline_packages, conflicts};
+}
+
+static System::ExitCodeAndOutput run_git_command(const VcpkgPaths& paths,
+                                                 const fs::path& working_directory,
+                                                 const std::string& cmd)
+{
+    const fs::path& git_exe = paths.get_tool_exe(Tools::GIT);
+    const fs::path dot_git_dir = working_directory / ".git";
+
+    const std::string full_cmd =
+        Strings::format(R"("%s" --work-tree="%s" %s)", fs::u8string(git_exe), fs::u8string(working_directory), cmd);
+
+    auto output = System::cmd_execute_and_capture_output(full_cmd);
+    return output;
+}
+
+const std::string get_version_commit_id(const std::string& package_name,
+                                        const std::string& requested_version,
+                                        const VcpkgPaths& paths)
+{
+    const auto database_filename = Strings::format("%s.db.json", package_name);
+    const auto database_file_path = paths.scripts / "port_versions_db" / database_filename;
+    Checks::check_exit(VCPKG_LINE_INFO,
+                       paths.get_filesystem().exists(database_file_path),
+                       "Version database file does not exist for package %s",
+                       package_name);
+
+    auto pair = Json::parse_file(VCPKG_LINE_INFO, paths.get_filesystem(), database_file_path);
+    Checks::check_exit(VCPKG_LINE_INFO, pair.first.is_object(), "Failed to parse %", database_filename);
+
+    auto& db_object = pair.first.object();
+
+    auto maybe_versions = db_object.get("versions");
+    Checks::check_exit(VCPKG_LINE_INFO,
+                       maybe_versions && maybe_versions->is_array(),
+                       "Database file %s contains no versions",
+                       database_filename);
+
+    auto& versions = maybe_versions->array();
+    for (auto&& version : versions)
+    {
+        auto version_string = version.object().get("version_string")->string().to_string();
+        if (version_string == requested_version)
+        {
+            return version.object().get("commit_id")->string().to_string();
+        }
+    }
+
+    Checks::exit_with_message(VCPKG_LINE_INFO, "Couldn't find version '%s' of %s", requested_version, package_name);
+}
+
+void Versions::fetch_port_versions(const VcpkgPaths& paths,
+                                   const ComputedVersions& versions,
+                                   const std::string& baseline)
+{
+    (void)baseline;
+    (void)versions;
+
+    Checks::check_exit(VCPKG_LINE_INFO, versions.conflicts.empty(), "There are conflicts in the computed versions.");
+
+    auto& fs = paths.get_filesystem();
+    const auto working_dir = paths.root / "vcpkg-temp";
```

Comment:
```suggestion
    const auto working_dir = paths.buildtrees / "versioning_tmp";
```
because underscores are not valid port names, this will avoid creating a new root folder and won't collide with ports.